### DISCUSSION TITLE
Securing use of third-party GitHub actions

### DISCUSSION THREAD

**Discussion by solarmosaic-kflorence** [community/community#18103] [Code Security]
Moving this discussion here: https://github.community/t/securing-use-of-third-party-github-actions/212989

---

In light of [GitHub's commitment to npm ecosystem security | The GitHub Blog 5](https://github.blog/2021-11-15-githubs-commitment-to-npm-ecosystem-security/) – I am wondering how GitHub plans to ensure the GitHub actions ecosystem also remains secure in the event of third-party action repository takeovers. As suggested in [Security hardening for GitHub Actions - GitHub Docs 13](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions) – the safest way at the moment is to pin any third-party actions to a commit SHA, but unfortunately most documentation on GitHub and in third-party actions suggest to pin to a major version (e.g. @v1), which means it will automatically pull in subsequent tags for that major version.

This is especially troubling for third-party actions which are invoked in an authenticated context. They could not only access the GITHUB_TOKEN, but it could also compromise cloud-accounts such as AWS if precautions are not taken.

These are problems that have already been solved in other ecosystems, through things like caching layers (to guarantee the resource does not change), multi-factor verification when publishing new versions of a package, and ensuring that previously published packages cannot be modified. At the very least, it seems that GitHub should promote security best practices in all places in their documentation, instead of expecting someone to find and read the security section. Even better if this could be enforced through a process or convention.

PS: this is even more of a nightmare when using an action that itself references other actions. The only way to get around that is to fork the action and maintain it yourself, or copy the action into your repository.

---

At the very least, providing some kind of manifest/lock file that could be defined in the .github folder, which would apply to all actions (even those referenced within actions), would be very useful. Currently we have to maintain commit SHAs across hundreds of workflow files in our monorepo and this would greatly improve that experience.

**Comment by L1ghtman2k** [comment]
A lock file equivalent would be helpful, especially for tracking transitive dependencies. I can use sha of the commit to pin an action, but it isn't always the case that that action also pins its dependencies.
